cudaextender/include/claraparabricks/genomeworks/cudaextender/extender.hpp

```diff
@@ -0,0 +1,103 @@
+/*
+* Copyright 2020 NVIDIA CORPORATION.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+#pragma once
+#include <claraparabricks/genomeworks/types.hpp>
+#include <claraparabricks/genomeworks/cudaextender/cudaextender.hpp>
+
+#include <vector>
+#include <memory>
+#include <cuda_runtime_api.h>
+
+namespace claraparabricks
+{
+namespace genomeworks
+{
+namespace cudaextender
+{
+
+struct SeedPair
+{
+    /// position of first sketch element in query_read_id_
+    position_in_read_t query_position_in_read;
+    /// position of second sketch element in target_read_id_
+    position_in_read_t target_position_in_read;
+};
+
+struct ScoredSegmentPair
+{
+    /// Seed for the segment
+    SeedPair seed_pair;
+    /// length of the segment
+    int32_t length;
+    /// score of the segment
+    int32_t score;
+
+    __host__ __device__ bool operator==(const ScoredSegmentPair& other) const
+    {
+        return ((seed_pair.target_position_in_read == other.seed_pair.target_position_in_read) && (seed_pair.query_position_in_read == other.seed_pair.query_position_in_read) && (length == other.length) && (score == other.score));
+    }
+};
+
+/// CUDA Extension object
+class Extender
+{
+public:
+    /// \brief Virtual destructor for Extender.
+    virtual ~Extender() = default;
+
+    /// \brief Host pointer prototype for extension
+    ///
+    /// Takes values from host data structures,
+    /// copies them over to device,
+    /// launches async extension kernels on specified stream. Filters
+    /// segments on device based on score_threshold
+    virtual StatusType extend_async(const char* h_query, int32_t query_length,
+                                    const char* h_target, int32_t target_length,
+                                    int32_t score_threshold,
+                                    std::vector<SeedPair>& h_seed_pairs) = 0;
+
+    /// \brief Device pointer prototype for  extension
+    ///
+    /// Memcopies to device memory are assumed to be done before this
+    /// function. Output array d_scored_segment_pairs must be pre-allocated on device.
+    /// Launches async extension kernel. Filters segments on device
+    /// based on score_threshold.
+    virtual StatusType extend_async(const char* d_query, int32_t query_length,
```

Comment:
@ahehn-nv @atadkase - we should also check with Sneha and Yatish if this API would make sense were we to implement the y drop algorithm

Reply:
For the y drop algorithm, a major change would be the output traceback pointers which forms the final alignment.

Reply:
Thanks @gsneha26 - this is super helpful! Since from an API point of view it's the trackback that's different, we can extend the ScoredSegmentPair struct or create one with alignment when the time comes to add drop